monai/data/meta_obj.py

```diff
@@ -0,0 +1,172 @@
+
+    This allows for subclassing `np.ndarray` and `torch.Tensor`.
+
+    Copying metadata:
+        * For `c = a + b`, then the meta data will be copied from the first
+        instance of `MetaImage`.
+    """
+
+    _meta: dict
+    _affine: torch.Tensor
+
+    def set_initial_val(self, attribute: str, input_arg: Any, input_tensor: MetaObj, default_fn: Callable) -> None:
+        """
+        Set the initial value. Try to use input argument, but if this is None
+        and there is a MetaImage input, then copy that. Failing both these two,
+        use a default value.
+        """
+        if input_arg is None:
+            input_arg = getattr(input_tensor, attribute, None)
+        if input_arg is None:
+            input_arg = default_fn(self)
+        setattr(self, attribute, input_arg)
+
+    @staticmethod
+    def get_tensors_or_arrays(args: Sequence[Any]) -> list[MetaObj]:
+        """
+        Recursively extract all instances of `MetaObj`.
+        Works for `torch.add(a, b)`, `torch.stack([a, b])` and numpy equivalents.
+        """
+        out = []
+        for a in args:
+            if isinstance(a, (list, tuple)):
+                out += MetaObj.get_tensors_or_arrays(a)
+            elif isinstance(a, MetaObj):
+                out.append(a)
+        return out
+
+    def _copy_attr(
+        self, attribute: str, input_objs: list[MetaObj], default_fn: Callable, deepcopy_required: bool
+    ) -> None:
+        """
+        Copy an attribute from the first in a list of `MetaObj`
+        In the cases `torch.add(a, b)` and `torch.add(input=a, other=b)`,
+        both `a` and `b` could be `MetaObj` or `torch.Tensor` so check
+        them all. Copy the first to the output, and make sure on correct
+        device.
+        Might have the MetaObj nested in list, e.g., `torch.stack([a, b])`.
+        """
+        attributes = [getattr(i, attribute) for i in input_objs]
+        if len(attributes) > 0:
+            val = attributes[0]
+            if deepcopy_required:
+                val = deepcopy(val)
+            if isinstance(self, torch.Tensor) and isinstance(val, torch.Tensor):
+                val = val.to(self.device)
+            setattr(self, attribute, val)
+        else:
+            setattr(self, attribute, default_fn())
+
+    def _copy_meta(self, input_meta_objs: list[MetaObj]) -> None:
+        """
+        Copy meta data from a list of `MetaObj`.
+        If there has been a change in `id` (e.g., `a+b`), then deepcopy. Else (e.g., `a+=1`), don't.
+        """
+        id_in = id(input_meta_objs[0]) if len(input_meta_objs) > 0 else None
+        deepcopy_required = id(self) != id_in
+        attributes = ("affine", "meta")
+        default_fns: tuple[Callable, ...] = (self.get_default_affine, self.get_default_meta)
+        for attribute, default_fn in zip(attributes, default_fns):
+            self._copy_attr(attribute, input_meta_objs, default_fn, deepcopy_required)
+
+    def get_default_meta(self) -> dict:
```

Comment:
It seemed cleaner to have specific functions for each of the default attributes. Otherwise there might be `{}` around the code but it wouldn't be obvious why.